### ISSUE TITLE
treesitter/highlighter.lua:145: end_col value outside range

### ISSUE THREAD

**Issue by entombedvirus** [neovim/neovim#12861]
- `nvim --version`: `NVIM v0.5.0-nightly-4-gbedab7e87`
- `vim -u DEFAULTS` (version: ) behaves differently? N/A
- Operating system/version: ubuntu-16.04
- Terminal name/version: kitty -> ssh -> tmux
- `$TERM`: tmux

### Steps to reproduce using `nvim -u NORC`

1. setup treesitter for go and enable TS syntax highlighting for go
2. open up the contents of https://gist.github.com/entombedvirus/19038b42b09439b227b0c6e75d977f5c in a .go file
3. jump to line 29, insert line above using `O`
4. type `x :` as if to type ` x := 1`

### Actual behaviour

As soon that `:` is pressed, the message area should now `Error executing lua callback: ...7e/share/nvim/runtime/lua/vim/treesitter/highlighter.lua:145: end_col value outside range` while still in `INSERT` mode.

NOTE: The error doesn't repro at different files or even in different lines in the same file. But I have 100% repro on that specific line on that specific file.

I added some extra logging to `treesitter/highlighter.lua`:

```
      if hl then
        local status, err = pcall(function()
            a.nvim_buf_set_extmark(self.buf, ts_hs_ns, start_row, start_col, {
                end_col = end_col,
                end_line = end_row,
                hl_group = hl
            })
        end)
        if not status then
            print(vim.inspect{
                status = status,
                err = err,
                start_row = start_row,
                start_col = start_col,
                end_col = end_col,
                end_line = end_row,
                hl_group = hl
            })
        end
      end

```

and got the following output:

```
{                                                                                                                                                                                         
  end_col = 24,                                                                                                                                                                           
  end_line = 30,                                                                                                                                                                          
  err = "...7e/share/nvim/runtime/lua/vim/treesitter/highlighter.lua:146: end_col value outside range",                                                                                   
  hl_group = 310,                                                                                                                                                                         
  start_col = 31,                                                                                                                                                                         
  start_row = 27,                                                                                                                                                                         
  status = false                                                                                                                                                                          
}                                                                                                                                                                                         
{                                                                                                                                                                                         
  end_col = 4,                                                                                                                                                                            
  end_line = 30,                                                                                                                                                                          
  err = "...7e/share/nvim/runtime/lua/vim/treesitter/highlighter.lua:146: end_col value outside range",                                                                                   
  hl_group = 338,                                                                                                                                                                         
  start_col = 0,                                                                                                                                                                          
  start_row = 30,                                                                                                                                                                         
  status = false                                                                                                                                                                          
}                                                                                                                                                                                         
{                                                                                                                                                                                         
  end_col = 6,                                                                                                                                                                            
  end_line = 30,                                                                                                                                                                          
  err = "...7e/share/nvim/runtime/lua/vim/treesitter/highlighter.lua:146: col value outside range",                                                                                       
  hl_group = 312,                                                                                                                                                                         
  start_col = 5,                                                                                                                                                                          
  start_row = 30,                                                                                                                                                                         
  status = false                                                                                                                                                                          
}                                                                                                                                                                                         
{                                                                                                                                                                                         
  end_col = 7,                                                                                                                                                                            
  end_line = 30,                                                                                                                                                                          
  err = "...7e/share/nvim/runtime/lua/vim/treesitter/highlighter.lua:146: col value outside range",                                                                                       
  hl_group = 345,                                                                                                                                                                         
  start_col = 6,                                                                                                                                                                          
  start_row = 30,                                                                                                                                                                         
  status = false                                                                                                                                                                          
}   
```

### Expected behaviour

no error messages.

**Comment by vigoux** [comment]
Hey, thanks for the report ! I _think_ that this is actually related to  #12867, just leaving that there for the record.

**Comment by krkhan** [comment]
I couldn't repro with the gist but I do run into this issue all the time seemingly at random with go + treesitter. Currently at version `NVIM v0.5.0-847-gad5652724`.

**Comment by xylix** [comment]
Happened with a Python file that contained a syntax error. Currently at `v0.5.0-nightly-244-g6af64d384`

**Comment by cormacrelf** [comment]
This has been a consistent, many different filetype problem for the last 5 HEADs I've tried over the last month. Rust, shell, yaml, javascript, typescript, it happens to all of them, and moreover (nobody has mentioned this) it destroys the highlighting for that file, spraying random highlights all over the place like a toddler got into the Sharpie box. You can fix it with `:e` but then it comes back pretty quickly. You probably don't want to release 0.5.0 without fixing it.

**Comment by bfredl** [comment]
lets see if tj did anything with the bytes validator, otherwise i will take it back :D

**Comment by gbrlsnchs** [comment]
> it destroys the highlighting for that file, spraying random highlights all over the place

This indeed happens, usually when I format my code outside the editor (via CLI) and then focus it back.

**Comment by bfredl** [comment]
> usually when I format my code outside the editor (via CLI) 

This is a bit more specific, and thus testable and actionable :]

I am working on a byte edit debug tool but it is still a bit rough https://github.com/tjdevries/nsync.nvim/pull/1.
My goal is to produce something that looks almost like failing testcases that people can just submit as issues.

**Comment by cormacrelf** [comment]
Come to think of it `git checkout` might be responsible for a large number of the instances for me. I couldn't say whether it's all of them, but it could be. I have been swapping branches quite a lot.

**Comment by vigoux** [comment]
@BK1603 I think this is related to our `:checktime` thing, but I am not sure yet.

**Comment by bfredl** [comment]
Hmm reloading a buffer _used_ to detach it, thus should force tree-sitter to re-attach and re-parse from scratch. perhaps this chain of logic has been broken somewhere?

**Comment by vigoux** [comment]
I think that the path took by `checktime` and `edit` are not the same, thus causing that.

**Comment by kristijanhusak** [comment]
Can confirm that this is also happening for me in javascript and lua.
Is there a way to restart the treesitter without doing `:e` ?

**Comment by bfredl** [comment]
The `checktime` issue is fixed on latest master (and next nightly built tonight). Please remind me of other issues with a clear step-by-step reproduce :D

**Comment by kristijanhusak** [comment]
@bfredl I managed to reproduce it like this.

Having this lua content:
```lua
vim.cmd[[augroup numbertoggle]]
  vim.cmd [[autocmd!]]
  vim.cmd [[autocmd BufEnter,FocusGained,InsertLeave,WinEnter * if &nu | set rnu   | endif]]
  vim.cmd [[autocmd BufLeave,FocusLost,InsertEnter,WinLeave   * if &nu | set nornu | endif]]
vim.cmd[[augroup END]]
```
And when typing in this command:
```
:%s/]]/
```

I get the error:
![screenshot](https://i.imgur.com/jIRIwlr.png)

I know that's not a valid lua syntax, but i wouldn't expect error to show up like that.

**Comment by akinsho** [comment]
I think the above error relates to #12967, not sure if you have `inccommand` enabled in your config @kristijanhusak.

**Comment by kristijanhusak** [comment]
@akinsho Yeah I have it enabled, so it's probably that.

**Comment by bfredl** [comment]
@kristijanhusak Cool, I will write the test case and hopefully also fix it :]

**Comment by vigoux** [comment]
Hey, I think I shortcutted you.

Could you please try https://github.com/neovim/neovim/pull/13539 and se if it fixes your issue ?

**Comment by kristijanhusak** [comment]
@bfredl I found another way to reproduce this.

1. Edit some file that has treesitter enabled and that is part of some git repo (I tested with lua file). Leave neovim opened on that file.
2. Open new terminal, go to the repo folder, and remove the changes you just did with git (i did `git reset --hard origin/master`)
3. Go back to the first neovim, see same error happening.

Usually when I get these errors, it is when I focus neovim after doing something outside of it.

**Comment by bfredl** [comment]
@kristijanhusak is this latest neovim? because it sounds just like the error I just fixed.

**Comment by kristijanhusak** [comment]
@bfredl yeah, latest master.
```
NVIM v0.5.0-dev+1009-g0efc71313
Build type: Release
LuaJIT 2.1.0-beta3
```
![ts-issue](https://user-images.githubusercontent.com/1782860/103694714-b8917e00-4f9b-11eb-9c94-37ec6fe668de.gif)

**Comment by marcelarie** [comment]
This happens to me constantly too. My version is the NVIM v0.5.0-dev+1025-g0fad952a2.

**Comment by zachliu** [comment]
same on `NVIM v0.5.0-dev+1050-g78f0f00cd`

**Comment by esetnik** [comment]
Same on
```
NVIM v0.5.0-dev+nightly-324-g4d1fc167a
Build type: Release
LuaJIT 2.1.0-beta3
```

**Comment by kristijanhusak** [comment]
@bfredl Is there a way to reload the highlights for a buffer? Currently i'm fixing this issue by doing `:edit`, but that's not really working when I have unsaved changes in the buffer. I just want to set up autocmd that will reload highlights on FocusGained.

**Comment by vigoux** [comment]
What I do is a combination of both : `:write` then `:edit` which can be automated in a function

**Comment by marcelarie** [comment]
" Autoread.(editor)
`    map <leader>a :e<cr>`

**Comment by luisiacc** [comment]
It happens to me on python when I create a new line setting a couple of classs variables, if I remove smartindent and autoindent from vimrc then no error shows up.
version: v0.5.0-dev+1016-g0af5a56e4

**Comment by horta** [comment]
nvim --version: NVIM v0.5.0-dev+94cf7bb
Operating system/version: macOS 11.2
Terminal name/version: kitty
$TERM: xterm-256color

**Steps to reproduce using** 

Have the following config:

```lua
-- file: ~/.config/nvim/init.lua

vim.cmd('packadd paq-nvim')
local paq = require('paq-nvim').paq
paq {'savq/paq-nvim', opt = true}
paq {'neovim/nvim-lspconfig'}
paq {'nvim-treesitter/nvim-treesitter', hook = vim.fn['TSUpdate']}

local ts = require 'nvim-treesitter.configs'
ts.setup {ensure_installed = 'maintained', highlight = {enable = true}}

local lsp = require 'lspconfig'
lsp.clangd.setup {cmd={'/opt/homebrew/Cellar/llvm/11.0.1/bin/clangd'}}

vim.api.nvim_set_keymap('n', 'gf', '<cmd>lua vim.lsp.buf.formatting()<cr>', {noremap = true})
```

nvim edit the following file

```c
int main(void) {
  float a = log(0);
  bbbbbbbbbbbb(log(0));
  float ccccccccccc = 1;
}
```

as follows

![bug](https://user-images.githubusercontent.com/514522/107162369-742e3f00-699a-11eb-84a7-1bd516569ab9.gif)

I type `gf` to format the file and cause the error.

**Comment by vigoux** [comment]
@luisiacc oh that's interesting, I remember we had issues with smartindent before... Could you provide a minimal reproducing example?

@horta yup I think I remember this one too, but having your reproduction steps will definitely help, thanks!

**Comment by horta** [comment]
Treesitter is great, super fast. I would be using it despite that bug =P

**Comment by rafikdraoui** [comment]
https://github.com/neovim/neovim/pull/13895 seems to have fixed this for me, when testing with the steps described in https://github.com/neovim/neovim/issues/12967#issuecomment-774072640

I'll have to use it longer before I'm convinced it takes care of all edge cases, but I'm hopeful!

**Comment by horta** [comment]
I will check it with my bug.c file.

**Comment by kristijanhusak** [comment]
I confirmed that my issue from commenet https://github.com/neovim/neovim/issues/12861#issuecomment-754876035 is fixed. 
I ran into similar problem when doing undo after trying to reproduce the steps.

1. Edit some file that has treesitter enabled and that is part of some git repo (I tested with lua file). Leave neovim opened on that file.
2. Open new terminal, go to the repo folder, and remove the changes you just did with git (i did git reset --hard origin/master)
3. Go back to the first neovim, see that file now refreshed and that content you just added is properly removed. No errors at this point, this was now fixed.
4. Try to undo changes that you just removed, error appears

Here's the gif:
![ts-undo-issue](https://user-images.githubusercontent.com/1782860/107659518-7f79b700-6c87-11eb-81e3-25aff63200d0.gif)

**Comment by horta** [comment]
The bug (https://github.com/neovim/neovim/issues/12861#issuecomment-774786548) persists with HEAD tree-sitter + neovim (commit 1aec5ba...)

![bug](https://user-images.githubusercontent.com/514522/107660860-9ec11600-6c80-11eb-92ff-a7e37713c8ed.gif)

**Comment by bfredl** [comment]
how is `vim.lsp.buf.formatting()` implemented? by a sequence of `nvim_buf_set_text` ?

**Comment by doubleloop** [comment]
Issue described by @kristijanhusak can be reliably reproduced with smaller config and without git:

testrc.vim:
```vim
call plug#begin('~/.local/share/nvim/plugged')
Plug 'nvim-treesitter/nvim-treesitter'
call plug#end()
lua require('nvim-treesitter.configs').setup{highlight={enable = true}}
```

```sh
$ echo "print('boo')" > test.py && vim -u testrc.vim test.py
```
inside vim (reproduces manually, does not reproduce when sourced as script):
```vim
:norm yyp
:w
:!echo "print('boo')" > %
:undo            " bug is already here (note wrong highlight)
:undo
:redo
:redo
```

![vim](https://user-images.githubusercontent.com/2602750/107684641-4d764e00-6ca3-11eb-965e-de79549f635e.gif)

**Comment by mfussenegger** [comment]
> how is `vim.lsp.buf.formatting()` implemented? by a sequence of `nvim_buf_set_text` ?

It uses `vim.lsp.util.apply_text_edits` which under the hood currently uses `nvim_buf_set_lines` but there is [a PR](https://github.com/neovim/neovim/pull/13703) to change it to use `nvim_buf_set_text`

**Comment by doubleloop** [comment]
https://github.com/neovim/neovim/pull/13935 fixed error msg after undo/redo :D Still I noticed some highlighting issue (described in https://github.com/neovim/neovim/issues/13937)

**Comment by bfredl** [comment]
not sure if noted elsewhere but diffput seems to fail sometimes.

**Comment by kristijanhusak** [comment]
From recent builds neovim started throwing some of these errors in the command line. For example I get this one a lot sometimes:

```
Error executing vim.schedule lua callback: /usr/local/share/nvim/runtime/lua/vim/highlight.lua:37: Line number outside range
```

I'm not exactly sure how to reproduce it yet.

**Comment by rexagod** [comment]
+1 @kristijanhusak I get this error as well atleast twice daily. I seems to occur when operating on a large chunk of code at once.

**Comment by xylix** [comment]
I'm pretty sure this happens to me with python and js when switching git branches while having existing neovim sessions open and then they reload the content. Altough I think recently this problem has mildened a bit; before the entire treesitter highlighting got disabled and there were `col value outside range` errors, now the errors still appear but some of the highlighting still works.

**Comment by bfredl** [comment]
@kristijanhusak that's probably not tree-sitter tho. We need to enable better tracebacks for `vim.schedule` so we can see what code invokes `vim.highlight` here

**Comment by kristijanhusak** [comment]
@bfredl is there a way to debug it?

**Comment by kristijanhusak** [comment]
@bfredl I figured it out. It was caused by efm-langserver returning `-1` for line range for diagnostics in files that are ignored (efm + eslint + ignored file with eslint). Switching to diagnosticls fixes the issue. Sorry for bothering.

**Comment by sarahkittyy** [comment]
Here is how I am able to reproduce it, on my setup. Apologies if it doesn't reproduce on anyone else's.

`nvim --version`: `NVIM v0.5.0-dev+1157-g0ab88c2ea`
I have the `nvim-treesitter` plugin, with the `cpp` parser installed and up-to-date.
`tree-sitter -V`: `tree-sitter 0.19.3`

[Here's one of the couple files that the bug occurs on for me](https://gist.github.com/sarahkittyy/1ba1abe2ba053a11d8322d25c73d6a01)

It's pulled from a random project of mine, I didn't have it in me to search through and try to hunt for the most minimal working example as it seems to happen on some files and not others.

Steps to reproduce:
- Open the file.
- Go to line 39 (maybe it works on any line, I'm just trying to be specific)
- Enter insert mode, hit `TAB`

Immediately, it breaks:
![image](https://user-images.githubusercontent.com/31551546/111738989-a9129900-8858-11eb-98ae-aa5f94f7f31c.png)

Here's hoping someone else can reproduce this. :D

**Comment by bfredl** [comment]
@sarahkittyy cannot reproduce. Perhaps options are important. What is your values of expandtab, tabstop, softtabstop and shiftwidth?

**Comment by sarahkittyy** [comment]
@bfredl noexpandtab, tabstop=4, softtabstop=4, shiftwidth=4

**Comment by chentau** [comment]
@sarahkittyy  see #14124 - I believe this pr should fix your issue.

**Comment by SandWoodJones** [comment]
This happens to me when editing rust files, but only after inserting a tab character. It was difficult to pin point the error as most of the error message was inline and didn't fit my terminal screen, so I couldn't read what the error code actually was.

**Comment by chentau** [comment]
@SandWoodJones Are you using `noexpandtab`?

**Comment by SandWoodJones** [comment]
yes

**Comment by chentau** [comment]
@SandWoodJones  Then can you try #14124 to see if that fixes the issue?

**Comment by ZymoticB** [comment]
Given a file in this state (I have the go treesitter parser installed from the grammar, and a gopls based lsp setup)
```
package main

import (

)

type test struct {
	a string
}

func main() {
	fmt.Println("Hello World")
}

func newTest() *test {
	return &test{
		a: "test",
	}
}
```
With `noexpandtab` if I go to column 0 of line 4, enter insert mode and insert a tab `i<tab>`, I see a `end_col value outside range` error on line 18.

With `expandtab` I do not get that error.

With #14124 rebased on master (because of the tree-sitter ABI change) and `noexpandtab` I do not get that error.

I do however continue to see that error in my real development flows but I am unable to create a minimal reproduction. In those cases, I cannot reproduce if I disable my `gopls` based language server. In those cases, I usually see the error by adding an extra tab to an empty line, exiting insert mode, and deleting the line. For example if I have

```
func f() {
<tab>fmt.Println("hello world")
}
```
and I modify this to
```
func f() {
<tab><tab>
<tab>fmt.Println("hello world")
}
```

and then delete the line with two tabs, I see the error.

**Comment by chentau** [comment]
@ZymoticB  Just to confirm, you don't see errors when your language server is disabled, correct? Are you using any auto-formatting from your language server?

**Comment by ZymoticB** [comment]
@chentau If I disable my language server I cannot reproduce, that is correct.

I'm honestly pretty new to a lot these pieces in neovim so I'm not sure if my language server is automatically formatting / providing indent information. I don't think it is though, I have a keybinding to `lua vim.lsp.buf.formatting()` that I use to explicitly format a file.

**Comment by ZymoticB** [comment]
Oh, its worth mentioning that I do have indent enabled for tree-sitter, however, I am able to reproduce the errors with tree-sitter indent disabled.

**Comment by fcomoreira1** [comment]
I am still having this error randomly, when undoing multiple times and redoing back to the normal state in a cpp file. I see that there is some movement to fix this issue, but just so I know, should this already be fixed in the nightly build?

**Comment by chentau** [comment]
@fcomoreira1 It's generally more helpful/relevant to know exactly what edits you were doing before the undo/redo, since in this case, it is most likely not undo or redo itself that causes an invalid update, but rather the undoing/redoing of a specific editing command (maybe `dd`, or `<c-v>5jc`. It would be great if you could provide that.

**Comment by chentau** [comment]
The `luado` problem should (hopefully) be resolved by #14318

**Comment by marcelbeumer** [comment]
Have the same issue (often not always) using `deno-fmt` on my TS code with `efm-language-server`. 

efm setup:
```
  local deno_fmt = {formatCommand = "deno fmt -", formatStdin = true}
  local lua_fmt = {formatCommand = "lua-format -i", formatStdin = true}

  nvim_lsp.efm.setup {
    on_attach = on_attach_common,
    init_options = {documentFormatting = true},
    filetypes = {"typescript", "typescriptreact" },
    settings = {
      rootMarkers = {".git/"},
      languages = {
        lua = { lua_fmt },
        typescript = { deno_fmt },
        typescriptreact = { deno_fmt },
      }
    }
  }
```

then running either `:lua vim.lsp.buf.formatting()` or `:lua vim.lsp.buf.formatting_sync({}, 1000)` manually.

LSPInfo looks all fine and green. Tried to check some debugging info using 
```
lua vim.lsp.set_log_level("debug")
lua vim.cmd('e'..vim.lsp.get_log_path())
```

But not sure what to look for?

Screenshot error:

<img width="1268" alt="image" src="https://user-images.githubusercontent.com/111366/115971684-1a292880-a54a-11eb-8615-2ef6d64aeea9.png">


Let me know if there's something specific I could do or check!

**Comment by artarf** [comment]
I had the same problem. It occured with my custom (lua) version of [parinfer_rust](https://github.com/eraserhd/parinfer-rust) which is kind of a autoformatter. Formatting is done after each modification. At first, my custom plugin used `nvim_buf_set_lines()` to set the whole buffer whenever something changed. When I changed it to set only changed rows, the error discussed in this issue started happening annoyingly often.

I changed the plugin to use `nvim_buf_set_text()` instead of `nvim_buf_set_lines()` and errors dissappeared (the area was exactly the same). I also tried `vim.fn.setline()` and no errors.

Wild guess: Using `nvim_buf_set_lines()` sometimes causes invalid change object to be sent to `treesitter`. That function is propably used in numerous plugins.

**Comment by FallenAngel97** [comment]
I also experience similar problem. 
![image](https://user-images.githubusercontent.com/14362650/117194845-4156e400-aded-11eb-9203-88b53e5ff066.png)

Example dart code:

```
import 'package:flutter/material.dart';

class ExampleClass extends StatefulWidget {
  @override
  State<StatefulWidget> createState() => ExampleClassState();
}

class ExampleClassState extends State<ExampleClass> {
  @override
  Widget build(BuildContext buildContext) {
    return Container();
  }
}

```

Start by typing "setState" and then press <CR> to confirm


My init.vim:

```
set number
set exrc
syntax on
set updatetime=300
set mouse=a
set nowrap
set hidden
filetype plugin indent on
set signcolumn=yes:2
call plug#begin('~/.local/share/nvim/plugged')
Plug 'nvim-treesitter/nvim-treesitter', {'do': ':TSUpdate'}
Plug 'sainnhe/sonokai' 
Plug 'neovim/nvim-lspconfig'
Plug 'nvim-lua/popup.nvim'
Plug 'nvim-lua/plenary.nvim'
Plug 'nvim-telescope/telescope.nvim'
Plug 'hrsh7th/nvim-compe'
Plug 'glepnir/galaxyline.nvim' , {'branch': 'main'}
Plug 'akinsho/flutter-tools.nvim'
Plug 'sbdchd/neoformat'
Plug 'kyazdani42/nvim-tree.lua'
call plug#end()
set wildmenu
set wildignore+=*.bmp,*.gif,*.ico,*.jpg,*.png,*.ico
set wildignore+=*.pdf,*.psd
set wildignore+=node_modules/*,bower_components/*
set cc=80
let g:gitgutter_highlight_linenrs = 1
set inccommand=nosplit
set wildmode=longest:full,full
let g:gitgutter_sign_added=''
let g:gitgutter_sign_removed=''
let g:gitgutter_sign_modified=''
set background=dark
colorscheme sonokai
nnoremap <Leader>t <cmd>Telescope find_files<CR>
nnoremap <Leader>f <cmd>Telescope live_grep<CR>
inoremap <silent><expr> <C-Space> compe#complete()
inoremap <silent><expr> <CR>      compe#confirm('<CR>')
nnoremap gd :lua vim.lsp.buf.definition()<CR>
autocmd BufWritePre *.dart lua vim.lsp.buf.formatting_sync(nil, 1000)
lua << EOF
vim.o.completeopt = "menuone,noselect"
require('telescope').setup{
   defaults = {
       prompt_position = 'top',
       sorting_strategy = 'ascending',
   }
}
require'compe'.setup {
  enabled = true;
  autocomplete = true;
  debug = false;
  min_length = 1;
  preselect = 'enable';
  throttle_time = 80;
  source_timeout = 200;
  incomplete_delay = 400;
  max_abbr_width = 100;
  max_kind_width = 100;
  max_menu_width = 100;
  documentation = true;

  source = {
    path = true;
    buffer = true;
    calc = true;
    nvim_lsp = true;
    nvim_lua = true;
    vsnip = true;
  };
}

require("flutter-tools").setup{
    flutter_path='/opt/flutter/bin/flutter',
    widget_guides = {
        enabled = true,
    },
    lsp = {
        settings = {
	    enableSdkFormatter=true,
	},
    },
}

require'nvim-treesitter.configs'.setup {
    highlight = {
        enable = true,
    },
    indent = {
       enable = true
    }
}
require("telescope").load_extension("flutter")
EOF
luafile ~/.config/nvim/eviline.lua
let g:nvim_tree_auto_close = 1
function! NvimTreeFindToggle()
lua << EOF
    local nt_lib = require'nvim-tree.view'
    if nt_lib.win_open() then
        nt_lib.close()
    else
        require'nvim-tree'.find_file(true)
    end
EOF
endfunction
nnoremap <C-b> :call NvimTreeFindToggle()<CR>

```

**Comment by chentau** [comment]
Can people experiencing issues with treesitter and LSP please try out #14501 and let me know if that PR helps out at all? It's a bit difficult for me to test LSP issues, since I don't use a language server.

**Comment by marcelbeumer** [comment]
Hi @chentau I just tried to your branch (e0a5eb732, chentau/extmark_setlines) and there I can not reproduce the issue anymore. Maybe still possible but I couldn't, switching back to `origin master` (516551319) I can reproduce it again.

**Comment by DarkDefender** [comment]
Even with all the latest fixes I'm still getting this error sometimes, though it happens very rarely now, so we are making progress!

Sadly I haven't run into any easy to reproduce cases yet.

**Comment by chentau** [comment]
Lsp issues should pretty much be fixed in latest master (sorry for breaking packer), but bug reports will still be very helpful. I think the remaining issues are probably edge cases that will be difficult to track down.

**Comment by kaddkaka** [comment]
I think it should be possible to do random testing here. With csmith for
example, if solar issue has been seen with clangd lsp?

Den ons 12 maj 2021 07:35Tony Chen ***@***.***> skrev:

**Comment by windwp** [comment]
Ok after I test a latest version on master .the error of virtual text is gone but the highlight is gone too :)) .
It display a white text.

**Comment by d3fkon** [comment]
It happens at least once an hour now. Cannot reproduce the steps, but I can confirm that 
1. Writing the file
2. Deleting the buffer `:bdelete`
3. Re opening the file
Seems to be my fix till we see it actually get fixed

**Comment by marcelbeumer** [comment]
@d3fkon what works for me is deleting the line that shows the treesitter error, and then undo

**Comment by sQVe** [comment]
@d3fkon @marcelbeumer The easiest way  to repair the broken state is by running `:e`.

**Comment by vedran314** [comment]
I was having this issue while using prettier/vim-prettier plugin all the time while writing ts and tsx files.

Switching to https://github.com/sbdchd/neoformat did the trick.

Just FYI

**Comment by vigoux** [comment]
Interesting, then we should compare those two plugins to see how they differ.
It may be something like a command that causes the problem.

**Comment by cormacrelf** [comment]
I have been getting this a lot using vimdiff, via tpope's vim-fugitive / `:Gdiffsplit`. Making many kinds of small edits to file in diff view causes the error. It's using the vimdiff folding strategy.

**Comment by dakahn** [comment]
On latest unstable and still seeing this issue with Treesitter + prettier in JS files. Can reproduce by opening a .js file. Making a change that needs formatting and hitting `<leader>p` to format which successfully formats the file but causes this error and broken highlighting
![image](https://user-images.githubusercontent.com/40970507/121737763-e6d13780-cabe-11eb-9422-55e6167adff6.png)

My uninteresting but potentially helpful Treesitter config copied right off the README
![image](https://user-images.githubusercontent.com/40970507/121737861-0c5e4100-cabf-11eb-9ec6-e2c020dd23da.png)

**Comment by adamansky** [comment]
I have the same problem in C code (CoC) randomly and quite often. In most cases code reformat or `dd` randomly cause what. Pretty sure the race condition here.

**Comment by tomodachijanai** [comment]
I found a reproducable case on latest nightly:
```
NVIM v0.5.0-dev+1398-gd09b8ad71
Build type: RelWithDebInfo
LuaJIT 2.1.0-beta3
Compilation: /usr/bin/cc -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -O2 -g -Og -g -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wimplicit-fallt
hrough -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=always -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DMIN_LOG_LEVEL=3 -I/home/runner/wor
k/neovim/neovim/build/config -I/home/runner/work/neovim/neovim/src -I/home/runner/work/neovim/neovim/.deps/usr/include -I/usr/include -I/home/runner/work/neovim/neovim/build/src/nvim/auto -I/home/runner/work/neovi
m/neovim/build/include
Compiled by runner@fv-az32-300

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/home/runner/work/neovim/neovim/build/nvim.AppDir/usr/share/nvim"

Run :checkhealth for more info
```

1. Clone https://github.com/nvim-telescope/telescope-fzf-native.nvim
2. Open file `test/test.c` on line 392
3. Run this macro `Vygv:s/char/int/^Mgvp` (repcale `^M` with an actual character for `<CR>` or just type this sequence manually)
![image](https://user-images.githubusercontent.com/45037206/121786381-2fc0e300-cbc8-11eb-99d7-becd3068452f.png)

**Comment by adamansky** [comment]
Current nightly version seems to be stable, almost two days I don't receive these errors anymore

**Comment by akinsho** [comment]
I'm definitely still getting them on a current nightly (sorry to be contradictory, just don't want this to get marked as resolved 😄). Primarily around deleting lines but I don't have a minimal reproduction yet.

**Comment by adamansky** [comment]
Hi guys! Sometimes this bug makes me bored a bit so I've made the investigation and gonna 
do some steps to resolve this issue. Let me briefly explain my investigation trip and propose
possible solution.

As mentioned above this bug is due to improper synchronization of treesitter tree state and nvim buffer state.

I have a good  example code which reproduces this bug with high probability - a simple 
svelte code snippet with prettier formatted. Almost 1/2 code formats ended with the error.

Heres is the code:
```svelte
<script lang="ts">
  import env from './env';
  const { content } = env;
</script>

<template>
  <div class="pros-block offset_120"> <!-- add/remove space before '>' then reformat -->
    <div class="pros-block__left dots__wrapper ">
      <img style="width: 100%;" src="images/Persons.png" alt="persons" />
      <img class="dots dots_left-bottom not-mobile" src="images/Dots.png" alt="dots" />
    </div>
    <div class="pros-block__right">
      <h1 class="title">{env.get($content, '/home/second/title')}</h1>
      <p class="text text_grey offset_20">{env.get($content, '/home/second/content')}</p>
    </div>
  </div>
  <div class="pros-section">
    <img src="images/Persons.png" alt="persons" />
  </div>
</template>
```

The  `prettier` code which modifies nvim buffer and applys formatted text:

```vim
autoload/prettier/utils/buffer.vim:

  " delete all lines on the current buffer
  silent! execute 'lockmarks %delete _'

  " replace all lines from the current buffer with output from prettier
  let l:idx = 0
  for l:line in l:newBuffer
    silent! lockmarks call append(l:idx, l:line)
    let l:idx += 1
  endfor

  " delete trailing newline introduced by the above append procedure
  silent! lockmarks execute '$delete _'
```

Then I located the place where treesitter tree receives updates (along with several other trivial places where 
    tree:parse() called):

https://github.com/neovim/neovim/blob/master/runtime/lua/vim/treesitter/languagetree.lua#L415 
```lua
function LanguageTree:_on_bytes(bufnr, changed_tick,
                          start_row, start_col, start_byte,
                          old_row, old_col, old_byte,
                          new_row, new_col, new_byte)
  self:invalidate()

  local old_end_col = old_col + ((old_row == 0) and start_col or 0)
  local new_end_col = new_col + ((new_row == 0) and start_col or 0)

  -- Edit all trees recursively, together BEFORE emitting a bytes callback.
  -- In most cases this callback should only be called from the root tree.
  self:for_each_tree(function(tree)
    tree:edit(start_byte,start_byte+old_byte,start_byte+new_byte,
      start_row, start_col,
      start_row+old_row, old_end_col,
      start_row+new_row, new_end_col)
  end)
...
```

`on_bytes` events produced by `nvim_buffer_attach()` in https://github.com/neovim/neovim/blob/master/runtime/lua/vim/treesitter.lua#L62 

```lua
  ...
  a.nvim_buf_attach(self:source(), false, {on_bytes=bytes_cb, on_detach=detach_cb, on_reload=reload_cb, preview=true})
```

**Looking ahead the following causes of this bug:**

1. `on_bytes` event fired not in all edit cases (some kind of evens never fired)
1. `on_bytes` event data for some cases is incorrect (I will show below)

Especially  for hurry guys here is the **dumb and durty workaround** which will work in most (but not all) cases:

```diff
--- a/runtime/lua/vim/treesitter/languagetree.lua
+++ b/runtime/lua/vim/treesitter/languagetree.lua
@@ -416,20 +416,7 @@ function LanguageTree:_on_bytes(bufnr, changed_tick,
                           start_row, start_col, start_byte,
                           old_row, old_col, old_byte,
                           new_row, new_col, new_byte)
-  self:invalidate()
-
-  local old_end_col = old_col + ((old_row == 0) and start_col or 0)
-  local new_end_col = new_col + ((new_row == 0) and start_col or 0)
-
-  -- Edit all trees recursively, together BEFORE emitting a bytes callback.
-  -- In most cases this callback should only be called from the root tree.
-  self:for_each_tree(function(tree)
-    tree:edit(start_byte,start_byte+old_byte,start_byte+new_byte,
-      start_row, start_col,
-      start_row+old_row, old_end_col,
-      start_row+new_row, new_end_col)
-  end)
-
+  self:invalidate(true)
   self:_do_callback('bytes', bufnr, changed_tick,
       start_row, start_col, start_byte,
       old_row, old_col, old_byte,
```

Returning to my svelte example, I logged `on_bytes` events during code format and found the following 
sequence of changes:

```
start_row=6 start_col=36 start_byte=131
old_row=0 old_col=0 old_byte=0
new_row=0 new_col=1 new_byte=1

start_row=6 start_col=36 start_byte=131
old_row=0 old_col=1 old_byte=1
new_row=0 new_col=0 new_byte=0

start_row=20 start_col=0 start_byte=673
old_row=1 old_col=0 old_byte=675
new_row=0 new_col=0 new_byte=0
```

But wait! There is `20 buffer append` events? 
```
  for l:line in l:newBuffer
    silent! lockmarks call append(l:idx, l:line)
```  

and why the last line edit (`$delete _`) has `old_byte=675`? it obviously should 
have value `old_byte=1`. 

```
  silent! lockmarks execute '$delete _'
```  

So here are signs of core problem what require some code refactoring:

1. `on_bytes` buffer notifications is wrongly coupled with `extmarks` code and for example when you do modifications
    with `lockmarks` option - no events will be fired. Curiosly `lockmarks` has not effect for the last `$delete _` command 
    (I've debugged and this flag gone away for some reason)
2. We have a problem with the last line deletion events (`old_byte=675`) which caused 
   by the following line https://github.com/neovim/neovim/blob/master/src/nvim/extmark.c#L542
   replacing it with `old_byte = (bcount_t)buf->deleted_bytes;` resolves this issue.
   Currently I have quite vague understanding of  meaning of `delete_bytes2` field.

## My proposition  

Move `on_bytes,..` staff away from extmarks closer to the buffer state, may be `change.c` will be good place for it.
and fix here missing `on_bytes` events and check again events data consistency. 

I may start preparing required changes in 1-2 weeks but help from project veterans will be invaluable here =)

**Comment by bfredl** [comment]
> Move on_bytes,.. staff away from extmarks closer to the buffer state, may be change.c will be good place for it.
and fix here missing on_bytes events and check again events data consistency.

Extmarks are already as close to"buffer state you can get. instead what we need to do is have `lockmarks` not affect extmark code at all.

> I may start preparing required changes in 1-2 weeks but help from project veterans will be invaluable here =)

go ahead, don't be afraid to ping us for any question :+1:

**Comment by adamansky** [comment]
Fix and test case landed waiting for review

**Comment by alexxGmZ** [comment]
Hi, I'm currently experiencing this problem in 0.10.

```
NVIM v0.10.0
Build type: Release
LuaJIT 2.1.1713484068
```

I encounter this problem when I use tabs for indention.

https://github.com/neovim/neovim/assets/77670158/c06ac7df-3f70-487b-95d3-6c3c7f29aee6

**Comment by clason** [comment]
Open a new issue with _complete_ and minimal reproduction steps.
